cbmc/proofs/proof_guide.md

```diff
@@ -0,0 +1,452 @@
+
+# Supply top-level contracts
+
+Now for the tricky bit. Does `XXX` need any top-level pre- and post-condition contracts, other that those inferred by its type-signature?
+
+A common case is where a function has an array parameter, where the numeric range of the array's elements are not just the range of the underlying predefined type.
+
+Similarly, a parameter of a predefined integer type might have to be constrained to a small range.
+
+Check for any BOUND macros in the body that give range constraints on the parameters, and make sure that your contracts are identical or consistent with those. Also read the comments for XXX to make sure they agree with your contracts.
+
+The order of the top-level contracts for a function is _always_:
+```
+t1 XXX(params...)
+    // clang-format off
+REQUIRES()
+ASSIGNS()
+ENSURES();
+// clang-format on
+```
+with a final semi-colon on the end of the last one. That final semi-colon is needed to terminate the function declaration as per normal C syntax.
+
+### Interior contracts and loop invariants
+
+If XXX contains no loop statements, then you might be able to just skip this step.
+
+As above, it's best if a function contains at most ONE top-level loop statement. Refer to the patterns above for common invariants. Don't be afraid to ask for help.
+
+### Prove it!
+
+Proof of a single function can be run from the proof directory for that function with `make result`.
+
+This produces `logs/result.txt` in plaintext format.
+
+Before pushing a new proof for a new function, make sure that _all_ proofs run OK from the `cbmc/proofs` directory with
+
+```
+export MLKEM_K=3 # or 2 or 4...
+./run-cbmc-proofs.py --summarize -j$(nproc)
+```
+
+That will use `$(nproc)` processor cores to run the proofs.
+
+### Debugging a proof
+
+If a proof fails, you can run
+
+```
+make result VERBOSE=1 >log.txt
+```
+
+and then inspect `log.txt` to see the exact sequence of commands that has been run. With that, you should be able to reproduce a failure on the command-line directly.
+
+## Worked Example - proving poly_tobytes()
+
+This section follows the recipe above, and adds actual settings, contracts and command to prove the `poly_tobytes()` function.
+
+### Populate a proof directory
+
+The proof directory is `cbmc/proofs/poly_tobytes`
+
+### Update Makefile
+
+The significant changes are:
+```
+HARNESS_FILE = poly_tobytes_harness
+PROOF_UID = poly_tobytes
+PROJECT_SOURCES += $(SRCDIR)/mlkem/poly.c
+CHECK_FUNCTION_CONTRACTS=$(MLKEM_NAMESPACE)poly_tobytes
+USE_FUNCTION_CONTRACTS=
+FUNCTION_NAME = $(MLKEM_NAMESPACE)poly_tobytes
+```
+Note that `USE_FUNCTION_CONTRACTS` is left empty since `poly_tobytes()` is a leaf function that does not call any other functions at all.
+
+### Update harness function
+
+`poly_tobytes()` has a simple API, requiring two parameters, so the harness function is:
+
+```
+void harness(void) {
+  poly a;
+  uint8_t r[MLKEM_POLYBYTES];
```

Comment:
Both of these seem to contradict the guidance you gave earlier in that pointer-type parameters should not use address-of-stack-allocated-object.